File: harness/tools/text_port_gate.py
"""harness/tools/text_port_gate.py -- emit the gate's cases, and diff the 6809 against the
reference. [T-P0-075 AC-3/AC-4]

★★★★★ THE PORT IS COMPARED TO THE PYTHON, AND THE PYTHON IS COMPARED TO THE ORACLE. The reference
leg is re-gated by harness/tools/text_run.sh on every run -- 4,594 rectangles and 293,648 glyphs
across nine titles -- so this leg cannot quietly rest on a stale reference. ★★★ Diffing the 6809
straight against the oracle would be the same comparison with a longer chain and one more thing to
get wrong; if this ever diverges, bisecting is one step [L-36].

★★★★ THE CASES ARE THE SWEEP'S, NOT A SELECTION. `--emit` walks agi.cpp's own rule -- every
loadable logic, its first eight non-empty texts, skipping the one message the pinned oracle cannot
survive -- which is byte for byte the list text_gate.py runs the reference over. **A gate whose
corpus is chosen for convenience is a claim about the convenience** [L-85].

★★★ §2P: the emitted case file carries game text, so it is written to build/ (gitignored) and
never committed, exactly as oracle_parser_cases.txt is. Everything this tool PRINTS is a count, a
position, a colour or a checksum.

usage:
  python harness/tools/text_port_gate.py --emit <game-dir> <cases-out> <table-out>
  python harness/tools/text_port_gate.py --check <game-dir> <results-in> [--fault-wrap]
"""
import argparse
import os
import struct
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "tools"))
sys.path.insert(0, os.path.dirname(__file__))
from volread import resource, logic as logic_mod                   # noqa: E402
from agivm import text as textref                                  # noqa: E402
from text_gate import _unsafe_for_oracle, MSGS_PER_LOGIC           # noqa: E402


def sweep_messages(game):
    """agi.cpp's sweep, and text_gate.py's reference walk, in one place."""
    for nr in range(256):
        try:
            raw = game.load("LOGIC", nr)
        except Exception:                                          # noqa: BLE001
            continue
        if not raw:
            continue
        lg = logic_mod.split(raw, nr)
        for msg in lg.messages[:MSGS_PER_LOGIC]:
            if not msg:
                continue
            if _unsafe_for_oracle(msg):
                continue
            yield nr, msg


def logic0_texts(game):
    try:
        raw0 = game.load("LOGIC", 0)
        if raw0:
            return logic_mod.split(raw0, 0).messages
    except Exception:                                              # noqa: BLE001
        pass
    return []


def do_emit(game_dir, cases_out, table_out):
    game = resource.load_from_files(game_dir)

    # ★★★ THE MESSAGE TABLE IS THE 6809's %g / %m SOURCE, and it is a pointer array into a text
    # block -- §2V's "a resource is in BANKED MEMORY, not a dict". The probe stages the block at a
    # known base and this writes the OFFSETS; the probe adds the base once.
    l0 = logic0_texts(game)
    blob = bytearray()
    offs = []
    for m in l0:
        if not m:
            offs.append(0xFFFF)          # ★ absent: the probe treats 0xFFFF as "no pointer"
            continue
        offs.append(len(blob))
        blob += m + b"\0"
    with open(table_out, "wb") as f:
        f.write(struct.pack(">H", len(offs)))
        for o in offs:
            f.write(struct.pack(">H", o))
        f.write(struct.pack(">H", len(blob)))
        f.write(bytes(blob))

    n = 0
    with open(cases_out, "wb") as f:
        for _nr, msg in sweep_messages(game):
            f.write(msg + b"\0")
            n += 1
    print("  emitted %d messages, logic0 table %d entries / %d text bytes"
          % (n, len(offs), len(blob)))
    return n


def reference_events(game_dir, fault_wrap=False, fault_printf=False):
    """The reference over the same list. Mirrors text_gate.py's own setup exactly."""
    textref.FAULT_WRAP = bool(fault_wrap)
    textref.FAULT_PRINTF = bool(fault_printf)
    game = resource.load_from_files(game_dir)
    l0 = [m.decode("latin-1") for m in logic0_texts(game)]
    st = textref.PrintfState(logic0_texts=l0, cur_logic_texts=l0)
    out = []
    for _nr, msg in sweep_messages(game):
        r = textref.TextRenderer(printf_state=st)
        r.checksum = reference_events.ck
        r.draw_message_box(msg.decode("latin-1"))
        m = r._msg
        glyphs = [e for e in r.events if e[0] == "G"]
        r.close_window()
        rect = r.events[-1]
        out.append({
            "boxw": m and 0 or 0,
            "rect": (rect[1], rect[2], rect[3], rect[4]),
            "glyphs": [(g[1], g[2], g[3], g[4], g[5]) for g in glyphs],
        })
        reference_events.ck = r.checksum
    textref.FAULT_WRAP = False
    textref.FAULT_PRINTF = False
    return out


reference_events.ck = 0


def read_results(path):
    """The probe's records. Glyphs precede their message header -- see text_probe.s."""
    data = open(path, "rb").read()
    msgs = []
    pend = []
    i = 0
    while i < len(data):
        t = data[i:i + 1]
        if t == b"G":
            row, col, fg, bg = data[i + 1], data[i + 2], data[i + 3], data[i + 4]
            ck = struct.unpack(">H", data[i + 5:i + 7])[0]
            pend.append((row, col, fg, bg, ck))
            i += 7
        elif t == b"M":
            boxw, boxh, srow, trow, tcol = data[i + 1:i + 6]
            bgx, bgy, bgw, bgh, ng = struct.unpack(">hhHHH", data[i + 6:i + 16])
            msgs.append({"rect": (bgx, bgy, bgw, bgh), "glyphs": pend[:ng],
                         "boxw": boxw, "boxh": boxh, "srow": srow,
                         "trow": trow, "tcol": tcol})
            pend = pend[ng:]
            i += 16
        else:
            raise SystemExit("★★★ bad record tag %r at offset %d" % (t, i))
    return msgs


def classify_rect(a, b):
    """★★★ L-103: the rectangle is DERIVED from the wrap, so it cannot name the stage that made
    it -- but WHICH FIELD moved narrows it. Same shape as text_gate.py's own classifier."""
    if a == b:
        return None
    if a[2] != b[2] and a[3] == b[3]:
        return "box WIDTH differs (textSize_Width)"
    if a[3] != b[3] and a[2] == b[2]:
        return "box HEIGHT differs (a different number of LINES)"
    if a[2] != b[2] and a[3] != b[3]:
        return "box WIDTH and HEIGHT both differ"
    return "POSITION differs (same size, different placement)"


def do_check(game_dir, results_in, fault_wrap, fault_printf):
    title = os.path.basename(os.path.normpath(game_dir))
    print("=== port gate: %s ===" % title)
    reference_events.ck = 0
    ours = read_results(results_in)
    theirs = reference_events(game_dir, fault_wrap, fault_printf)
    print("  messages     : 6809 %d   reference %d" % (len(ours), len(theirs)))
    og = sum(len(m["glyphs"]) for m in ours)
    tg = sum(len(m["glyphs"]) for m in theirs)
    print("  glyphs       : 6809 %d   reference %d" % (og, tg))

    ok = True
    n = min(len(ours), len(theirs))
    for i in range(n):
        why = classify_rect(ours[i]["rect"], theirs[i]["rect"])
        if why:
            print("  *** RECTANGLE %d DIVERGES -- %s" % (i, why))
            print("      6809 %s   reference %s" % (ours[i]["rect"], theirs[i]["rect"]))
            ok = False
            break
    if ok:
        print("  OK rectangles identical (%d)" % n)

    bad = 0
    for i in range(n):
        a, b = ours[i]["glyphs"], theirs[i]["glyphs"]
        if a != b:
            for j in range(min(len(a), len(b))):
                if a[j] != b[j]:
                    print("  *** GLYPH message %d index %d: 6809 %s   reference %s"
                          % (i, j, a[j], b[j]))
                    break
            else:
                print("  *** GLYPH COUNT message %d: 6809 %d   reference %d"
                      % (i, len(a), len(b)))
            bad += 1
            if bad >= 3:
                break
    if bad:
        ok = False
    elif ok:
        print("  OK glyphs identical in position, colour and checksum (%d)" % og)

    if len(ours) != len(theirs) or og != tg:
        ok = False
    print("PORT GATE %s" % ("PASS" if ok else "*** FAIL"))
    return 0 if ok else 1


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--emit", nargs=3, metavar=("GAMEDIR", "CASES", "TABLE"))
    ap.add_argument("--check", nargs=2, metavar=("GAMEDIR", "RESULTS"))
    ap.add_argument("--fault-wrap", action="store_true",
                    help="run the REFERENCE with its wrap fault, so a correct port must FAIL")
    ap.add_argument("--fault-printf", action="store_true",
                    help="run the REFERENCE with its printf fault, so a correct port must FAIL")
    a = ap.parse_args()
    if a.emit:
        do_emit(*a.emit)
        return 0
    if a.check:
        return do_check(a.check[0], a.check[1], a.fault_wrap, a.fault_printf)
    ap.print_help()
    return 2


if __name__ == "__main__":
    sys.exit(main())

File: harness/tools/text_port_gate.lua
-- harness/tools/text_port_gate.lua -- run the PORTED text engine over the gate's messages.
-- [T-P0-075 AC-3]
--
-- ★★★★ THE SAME MESSAGE LIST THE REFERENCE RAN. text_port_gate.py --emit writes the sweep's own
-- messages, NUL-terminated, plus logic 0's pointer table; this stages both byte for byte and
-- reads the probe's decision records back. The comparison is --check's, not this file's.
--
-- ★★★★★ FED IN CHUNKS, AND THE CHUNK IS BOUNDED AT BOTH ENDS. The parser gate learned this the
-- expensive way: a single-shot run had the probe reading bytes its own results had overwritten
-- from case 993 onward, and the symptom was a plausible wrong answer rather than a crash. Here
-- the MESSAGE area and the RESULT area are disjoint regions, and the probe additionally refuses a
-- message when fewer than 4,812 result bytes remain -- setting TX_ROVF so this file shortens the
-- chunk rather than discovering the truncation in the diff.
--
-- ★★ Environment: TXP_PROG, TXP_CASES, TXP_TABLE, TXP_OUT, TXP_CHUNK, TXP_MAP.
-- ★ §2P: stages message text into RAM, writes positions, colours and checksums. Nothing printed.

local m    = manager.machine
local cpu  = m.devices[":maincpu"]
local prog = cpu.spaces["program"]

local PROG  = os.getenv("TXP_PROG")  or "build/text_probe.bin"
local CASES = os.getenv("TXP_CASES") or "build/text_cases.bin"
local TABLE = os.getenv("TXP_TABLE") or "build/text_table.bin"
local OUT   = os.getenv("TXP_OUT")   or "build/text_6809_results.bin"
local CHUNK = tonumber(os.getenv("TXP_CHUNK") or "64")
local MAP   = os.getenv("TXP_MAP")   or "build/text_probe.map"

-- ★★★★★ SYMBOLS FROM THE MAP, NEVER FROM A LITERAL, and the map must be as new as the binary.
-- parser_gate.lua's header records what a stale hard-coded address cost: a diagnostic that
-- printed `fpos=33849` from three addresses that had gone two bytes stale, and four titles
-- investigated on the strength of a number the diagnostic invented.
local sym = {}
do
    local f = io.open(MAP, "r")
    if f then
        for line in f:lines() do
            local n, v = line:match("^Symbol:%s+(%S+)%s+%b()%s+=%s+(%x+)")
            if n then sym[n] = tonumber(v, 16) end
        end
        f:close()
    end
end
local function symbol(n)
    if not sym[n] then
        print(string.format("★★★ no symbol '%s' in %s -- assemble with --map", n, MAP))
        m:exit(); error("missing symbol " .. n)
    end
    return sym[n]
end

-- ★★★★ `lfs` IS NOT A GLOBAL IN MAME'S LUA -- only via require("lfs"). A guard that read the
-- global got nil, fell through to a "does the file exist" fallback, compared 0 < 0 and passed a
-- deliberately stalened map [P6.3]. This is the corrected form.
local lfs = select(2, pcall(require, "lfs"))
local function mtime(p)
    local ok, a = pcall(function() return lfs and lfs.attributes(p, "modification") end)
    if ok and a then return a end
    return nil
end

local function nearest(addr)
    local bn, bv = nil, -1
    for n, v in pairs(sym) do
        if v <= addr and v > bv then bn, bv = n, v end
    end
    if not bn then return "?" end
    return string.format("%s+%d", bn, addr - bv)
end

local TX_GO, TX_DONE, TX_NMSG, TX_RUN, TX_ROVF = 0x0020, 0x0021, 0x0022, 0x0024, 0x0026
local TX_TABLE, TX_MSGS   = 0x3000, 0x5000
local TX_MSGS_END         = 0x9000
local TX_RESULTS, TX_RES_END = 0x9400, 0xFE00

local function slurp(p)
    local f = io.open(p, "rb"); if not f then return nil end
    local d = f:read("*a"); f:close(); return d
end

local function poke(addr, data)
    for i = 1, #data do prog:write_u8(addr + i - 1, data:byte(i)) end
end

-- split the case blob into NUL-terminated messages, host-side
local cases = {}
do
    local blob = slurp(CASES)
    if not blob then print("★★★ no cases at " .. CASES); os.exit(1) end
    local i = 1
    while i <= #blob do
        local j = blob:find("\0", i, true)
        if not j then break end
        cases[#cases + 1] = blob:sub(i, j)      -- keep the terminator
        i = j + 1
    end
end

local outf = io.open(OUT, "wb")
local phase, cursor, staged_n, booted = "boot", 1, 0, false
local chunk = CHUNK

local function stage_chunk()
    local addr, n = TX_MSGS, 0
    while cursor + n <= #cases and n < chunk do
        local c = cases[cursor + n]
        if addr + #c > TX_MSGS_END then break end
        poke(addr, c)
        addr = addr + #c
        n = n + 1
    end
    staged_n = n
    prog:write_u8(TX_NMSG, (n >> 8) & 0xFF)
    prog:write_u8(TX_NMSG + 1, n & 0xFF)
    prog:write_u8(TX_ROVF, 0)
    prog:write_u8(TX_GO, 1)
end

local function read_chunk()
    -- ★★★ The probe stops early on an overflow and reports how many it actually did, so the
    -- number of records read is TX_RUN and not what was staged. A reader that trusted the staged
    -- count would read whatever followed the last record as a record.
    local ran = prog:read_u8(TX_RUN) * 256 + prog:read_u8(TX_RUN + 1)
    local ovf = prog:read_u8(TX_ROVF)
    local addr, done, bytes = TX_RESULTS, 0, {}
    while done < ran and addr < TX_RES_END do
        local tag = prog:read_u8(addr)
        if tag == 0x47 then                     -- 'G', 7 bytes
            for k = 0, 6 do bytes[#bytes + 1] = string.char(prog:read_u8(addr + k)) end
            addr = addr + 7
        elseif tag == 0x4D then                 -- 'M', 16 bytes
            for k = 0, 15 do bytes[#bytes + 1] = string.char(prog:read_u8(addr + k)) end
            addr = addr + 16
            done = done + 1
        else
            print(string.format("★★★ bad record tag $%02X at $%04X after %d messages",
                                tag, addr, done))
            m:exit(); return
        end
    end
    outf:write(table.concat(bytes))
    cursor = cursor + ran
    if ovf ~= 0 and ran < staged_n then
        -- shrink so the next attempt fits; never below 1, or the run cannot progress
        chunk = math.max(1, math.floor(chunk / 2))
        print(string.format("  result buffer full after %d of %d -- chunk now %d",
                            ran, staged_n, chunk))
    end
end

_G._tx = emu.add_machine_frame_notifier(function()
    if not booted then
        if m.time:as_double() < 0.3 then return end
        booted = true
        local blob = slurp(PROG)
        if not blob then print("★★★ no program at " .. PROG); m:exit(); return end
        local tb, tm = mtime(PROG), mtime(MAP)
        if not tm then
            print(string.format("★★★ cannot stat %s -- no symbol map, or lfs unavailable.", MAP))
            m:exit(); return
        end
        if tb and tm < tb then
            print(string.format("★★★ %s is OLDER than %s -- the map is stale and every symbol it "
                                .. "names is a guess. Re-assemble with --map.", MAP, PROG))
            m:exit(); return
        end
        poke(0x2000, blob)

        -- ── the logic-0 table: u16 count, u16 offsets, u16 textlen, then the text ──
        local t = slurp(TABLE)
        if not t then print("★★★ no table at " .. TABLE); m:exit(); return end
        local nent = t:byte(1) * 256 + t:byte(2)
        local tlen = t:byte(3 + nent * 2) * 256 + t:byte(4 + nent * 2)
        local textbase = TX_TABLE + 2 + nent * 2
        if textbase + tlen > TX_MSGS then
            print(string.format("★★★ logic-0 text block is %d bytes; %d available before $%04X",
                                tlen, TX_MSGS - textbase, TX_MSGS))
            m:exit(); return
        end
        -- ★★ Offsets become ABSOLUTE pointers here, once, rather than in the probe: the 6809's
        -- %g/%m path is then a table lookup and nothing else. 0xFFFF marks an absent message and
        -- becomes a null pointer, which text.s tests for.
        for i = 0, nent - 1 do
            local o = t:byte(3 + i * 2) * 256 + t:byte(4 + i * 2)
            local p = (o == 0xFFFF) and 0 or (textbase + o)
            prog:write_u8(TX_TABLE + i * 2, (p >> 8) & 0xFF)
            prog:write_u8(TX_TABLE + i * 2 + 1, p & 0xFF)
        end
        poke(textbase, t:sub(5 + nent * 2))

        local L0BASE, L0N = symbol("txt_l0base"), symbol("txt_l0n")
        local CURBASE, CURN = symbol("txt_curbase"), symbol("txt_curn")
        prog:write_u8(L0BASE, (TX_TABLE >> 8) & 0xFF)
        prog:write_u8(L0BASE + 1, TX_TABLE & 0xFF)
        prog:write_u8(CURBASE, (TX_TABLE >> 8) & 0xFF)
        prog:write_u8(CURBASE + 1, TX_TABLE & 0xFF)
        -- ★★★★ %m READS LOGIC 0 IN THE SWEEP, because the sweep never sets curLogicNr and so it is
        -- 0. Measured in P6.18 against the oracle's own rectangle for Kingquest1 message 555, and
        -- the reference carries the same setting -- so both legs point here.
        prog:write_u8(L0N, nent & 0xFF)
        prog:write_u8(CURN, nent & 0xFF)

        print(string.format("staged: program %d B, %d messages, logic0 %d entries / %d text B",
                            #blob, #cases, nent, tlen))
        cpu.state["PC"].value = 0x2000
        phase = "stage"
        return
    end

    if phase == "stage" then
        if cursor > #cases then
            outf:close()
            print(string.format("★ %d messages run -> %s", cursor - 1, OUT))
            m:exit(); return
        end
        stage_chunk()
        phase = "run"
        return
    end

    if phase == "run" then
        if prog:read_u8(TX_DONE) ~= 1 then
            -- ★★★★ THE BUDGET IS AGAINST PROGRESS, NOT AGAINST THE CHUNK. The parser gate declared
            -- a working chunk stalled because its clock started when the chunk did and never
            -- restarted; 224 cases fit and 254 did not, and four titles were reported as
            -- diverging. Reset whenever TX_RUN advances.
            local run_now = prog:read_u8(TX_RUN) * 256 + prog:read_u8(TX_RUN + 1)
            if _G._tx_run ~= run_now then _G._tx_run, _G._tx_t = run_now, m.time:as_double() end
            if not _G._tx_t then _G._tx_t = m.time:as_double() end
            if m.time:as_double() - _G._tx_t > 20 then
                local pc = cpu.state["PC"].value
                print(string.format("★★★ stalled: DONE=%d RUN=%d GO=%d PC=$%04X (%s) S=$%04X "
                                    .. "staged=%d cursor=%d",
                                    prog:read_u8(TX_DONE), run_now, prog:read_u8(TX_GO),
                                    pc, nearest(pc), cpu.state["S"].value, staged_n, cursor))
                -- ★★ Name the engine's own cursors: a stall in the wrap and one in the
                -- substitution look identical from outside, and they are different bugs.
                local SRC, DST, SP = symbol("txt_src"), symbol("txt_dst"), symbol("txt_sp")
                local CR, WS, BH = symbol("txt_cr"), symbol("txt_ws"), symbol("txt_boxh")
                print(string.format("     src=$%04X dst=$%04X sp=%d  cur_read=$%04X "
                                    .. "word_start=$%04X boxh=%d",
                                    prog:read_u8(SRC) * 256 + prog:read_u8(SRC + 1),
                                    prog:read_u8(DST) * 256 + prog:read_u8(DST + 1),
                                    prog:read_u8(SP),
                                    prog:read_u8(CR) * 256 + prog:read_u8(CR + 1),
                                    prog:read_u8(WS) * 256 + prog:read_u8(WS + 1),
                                    prog:read_u8(BH)))
                -- ★★★★★ AND THE RETURN CHAIN. A PC inside a DATA region means control was
                -- transferred there, and only two things do that: a `jsr ,x` with a bad X, or an
                -- `rts`/`puls pc` pulling data as an address. **The stack says which**, and
                -- without it the investigation is a reading of the wrap that cannot conclude.
                -- ★★★ Every candidate address is resolved through the map, so a frame that IS a
                -- return address is named and one that is not reads as an offset into a buffer.
                local s = cpu.state["S"].value
                local frames = {}
                for k = 0, 11 do
                    local v = prog:read_u8(s + k * 2) * 256 + prog:read_u8(s + k * 2 + 1)
                    frames[#frames + 1] = string.format("%04X(%s)", v, nearest(v))
                end
                print("     stack: " .. table.concat(frames, " "))
                local ib = {}
                for k = -4, 6 do
                    ib[#ib + 1] = string.format("%02X", prog:read_u8(pc + k))
                end
                print(string.format("     bytes at PC-4..PC+6: %s", table.concat(ib, " ")))
                local TOUT = symbol("tx_out")
                print(string.format("     tx_out=$%04X  txt_gcb=$%04X  txt_emit=$%04X",
                                    prog:read_u8(TOUT) * 256 + prog:read_u8(TOUT + 1),
                                    prog:read_u8(symbol("txt_gcb")) * 256
                                        + prog:read_u8(symbol("txt_gcb") + 1),
                                    prog:read_u8(symbol("txt_emit")) * 256
                                        + prog:read_u8(symbol("txt_emit") + 1)))
                m:exit()
            end
            return
        end
        _G._tx_t, _G._tx_run = nil, nil
        read_chunk()
        prog:write_u8(TX_GO, 0)
        phase = "ack"
        return
    end

    if phase == "ack" then
        if prog:read_u8(TX_DONE) ~= 0 then return end
        phase = "stage"
    end
end)
